Add assets page (#2779)
withastro/docs: src/content/docs/en/reference/image-service-reference.mdx

Services definitions take the shape of an exported default object with various required methods ("hooks") to create the required properties for your image. 

Local services have a `transform` method available to perform transformations on your image. External services can only return the final URL and image attributes.

### Local Service

```js
import type { LocalImageService } from "astro";

const service: LocalImageService = {
  getURL(options: ImageTransform) {
    return `/my_super_endpoint_that_transforms_images?w=${options.width}`;
  },
  parseURL(url: URL) {
    return {
      width: url.searchParams.get("w"),
    };
  },
  transform(options: ImageTransform) {
    const { buffer } = mySuperLibraryThatEncodesImages(options);

    return {
      data: buffer,
      format: options.format,
    };
  },
};

export default service;
```

### External Service

```js
import type { ExternalImageService } from "astro";

const service: ExternalImageService = {
  getURL(options: ImageTransform) {
    return `https://mysupercdn.com/${options.src}?q=${options.quality}`;
  },
};

export default service;
```

## Hooks

### `getURL()`

**Required for local and external services**

> `getURL(options: ImageTransform): string`

Reviewer: Here I am fine with leaving it as `ImageTransform` since it is explained below what `options` contain.

  - For local services, return the URL of the endpoint managing your transformation (in SSR and dev).
  - For external services, return the final URL of the image.
  - `options` contain the parameters passed by the user

### `parseURL():`

**Required for Local services; unavailable for external services**

> `parseURL(url: URL): ImageTransform`
>   - For SSR and dev, parses the generated URLs by `getURL` back into an ImageTransform to be used by `transform`.

Reviewer: ```suggestion
`parseURL(url: URL): { src: string, [key: string]: any}`
  - For SSR and dev, parses the generated URLs by `getURL` back into an object with the different properties to be used by `transform`.
```

### `transform()`

**Required for Local services only; unavailable for external services**

> `transform(buffer: Buffer, options: ImageTransform): { data: Buffer, format: OutputFormat }`

Reviewer: ```suggestion
`transform(buffer: Buffer, options: { src: string, [key: string]: any }): { data: Buffer, format: OutputFormat }`
```

  - Transform and return the image. It is necessary to return a `format` to ensure that the proper MIME type is served to users in development and SSR.

Ultimately, in development and SSR, it is up to the local endpoint (that `getURL` points to) to call both `parseURL` and `transform` if wanted. `transform` however, is called during the build to create the final assets files.

### `getHTMLAttributes()`

**Optional for both local and external services**

`getHTMLAttributes(options: ImageTransform): Record<string, any>`

>   - Return all additional attributes needed to render the image in HTML. For instance, you might want to return a specific `class` or `style`, or `width` and `height`.

Reviewer: ```suggestion
  - Return all additional attributes needed to render the image in HTML. For instance, you might want to return a specific `class` or `style`, or `width` and `height`.
  - `options` contain the parameters passed by the user
```

---
title: Image Service API
i18nReady: false
---

Astro [assets](/en/guides/assets/) were designed to make it easy for any image optimization service to build a service on top of Astro.

:::caution
The Image Services API is an _experimental_ Astro feature introduced in v2.1. This API is subject to change before it is marked as stable.
:::

## What is an Image Service?

Astro provides two types of image services: Local and External.

> - **Local services** handle image transformations directly at build for static sites, or at runtime in dev / SSR. These are often wrappers around libraries like Sharp, ImageMagick or Squoosh.
> - **External services** point to URLs and can add support for services such as Cloudinary, Vercel or any RIAPI-compliant server.

Reviewer 2: Serial commas
```suggestion
- **Local services** handle image transformations directly at build for static sites, or at runtime in dev / SSR. These are often wrappers around libraries like Sharp, ImageMagick, or Squoosh.
- **External services** point to URLs and can add support for services such as Cloudinary, Vercel, or any RIAPI-compliant server.
```

Local services have a `transform()` method available to perform transformations on your image. External services can only return the final URL and image attributes.

### Local Service

To create your own local service, you can choose to use the built-in endpoint (`/_image`) and pass your own `transform` function, or you can additionally create your own endpoint that calls the service's functions.

```js
import type { LocalImageService } from "astro";
const service: LocalImageService = {
  getURL(options: ImageTransform) {
    const href = typeof options.src === "string" ? options.src : options.src.src;
    return `/my_super_endpoint_that_transforms_images?href={href}&w=${options.width}`;
  },
  parseURL(url: URL) {
    return {
      src: url.searchParams.get("href"),
      width: url.searchParams.get("w"),
    };
  },
  transform(buffer: Buffer, options: { src: string, [key: string]: any }): { data: Buffer, format: OutputFormat } {
    const { buffer } = mySuperLibraryThatEncodesImages(options);
    return {
      data: buffer,
      format: options.format,
    };
  },
};
export default service;
```

#### How this Works

> At build time for static sites, both `<Image/>` and `getImage(options)` call the `transform()` function. They pass options either through component attributes or an `options` argument, respectively. The transformed images will be built to a `dist/_astro` folder.

Reviewer 3: ```suggestion
At build time for static sites and pre-rendered routes, both `<Image/>` and `getImage(options)` call the `transform()` function. They pass options either through component attributes or an `options` argument, respectively. The transformed images will be built to a `dist/_astro` folder.
```

- **Local services** handle image transformations directly at build for static sites, or at runtime both in development mode and SSR. These are often wrappers around libraries like Sharp, ImageMagick, or Squoosh. In dev mode and in SSR, local services use an API endpoint to do the transformation.
- **External services** point to URLs and can add support for services such as Cloudinary, Vercel, or any [RIAPI](https://github.com/riapi/riapi)-compliant server.

## Building using the Image Services API

Services definitions take the shape of an exported default object with various required methods ("hooks") to create the required properties for your image. 

Local services have a `transform()` method available to perform transformations on your image. External services can only return the final URL and image attributes.

### Local Service

To create your own local service, you can choose to use the built-in endpoint (`/_image`) and pass your own `transform` function, or you can additionally create your own endpoint that calls the service's functions.

```js
import type { LocalImageService } from "astro";
const service: LocalImageService = {
  getURL(options: ImageTransform) {

>     const src = typeof options.src === "string" ? options.src || options.src.src;
>     return `/my_super_endpoint_that_transforms_images?shref{srcString}&w=${options.width}`;

Reviewer: ```suggestion
    const src = typeof options.src === "string" ? options.src : options.src.src;
    return `/my_super_endpoint_that_transforms_images?href={src}&w=${options.width}`;
```

  },
  parseURL(url: URL) {
    return {
      src: url.searchParams.get("href"),
      width: url.searchParams.get("w"),
    };
  },
  transform(buffer: Buffer, options: { src: string, [key: string]: any }): { data: Buffer, format: OutputFormat } {
    const { buffer } = mySuperLibraryThatEncodesImages(options);
    return {
      data: buffer,
      format: options.format,
    };
  },
};
export default service;
```

#### How this Works

> At build time for static sites, both `<Image/>` and `getImage(options)` call the `transform()` function. They pass options either through component attributes or an `options` argument, respectively. The transformed images will be built to a `dist/_astro` folder.

Reviewer: This is also the case for pre-rendered routes, if you want to add an additional precision.

In dev mode and SSR mode, Astro doesn't know ahead of time which images need to be optimized. Astro uses a GET endpoint (by default, (`/_image`) to process the images at runtime. `<Image/>` and `getImage()` pass their options to `getURL()`, which will return the endpoint URL. Then, the endpoint calls `parseURL()` and passes the resulting properties to `transform()`.

#### getConfiguredImageService

### External Service

To use an external service . . . 

```js
import type { ExternalImageService } from "astro";

const service: ExternalImageService = {
  getURL(options: ImageTransform) {
    return `https://mysupercdn.com/${options.src}?q=${options.quality}`;
  },
};

export default service;
```

## Hooks

### `getURL()`

**Required for local and external services**

`getURL(options: ImageTransform): string`

> For local services, this hook returns the URL of the endpoint that generates your image (in SSR and dev mode). The local endpoint that `getURL()` points to must call both `parseURL()` and `transform()`.

Reviewer: Calling `parseURL` and `transform` is technically optional. You might have cases where your endpoint isn't written in JavaScript, and as such your service won't use `parseURL`

For external services, this hook returns the final URL of the image.

For both services, `options` are the properties passed by the user as attributes of the `<Image />` component or as options to `getImage()`. They are of the following type:

```ts
export type ImageTransform = {
    // ESM imported images | remote/public image paths
    src: ImageAsset | string;
    width?: number;
    height?: number;
    quality?: ImageQuality;
    format?: OutputFormat;
    alt?: string;
    [key: string]: any;
};
```

> ### `parseURL():`

Reviewer: ```suggestion
### `parseURL()`
```

---
title: Image Service API
i18nReady: false
---

Astro [Assets](/en/guides/assets) were designed to make it easy for any image optimization service to build a service on top of Astro.

> 

Reviewer 3: ```suggestion

::::caution
The Image Services API is an _experimental_ Astro feature introduced in v2.1. Be mindful that the API is subject to change before it is marked as stable.
:::

```

i18nReady: false
---

Astro [assets](/en/guides/assets/) were designed to make it easy for any image optimization service to build a service on top of Astro.

:::caution
The Image Services API is an _experimental_ Astro feature introduced in v2.1. Be mindful that the API is subject to change before it is marked as stable.
:::

## What is an Image Service?

Astro provides two types of image services: Local and External.

> - **Local services** handle image transformations directly at build for static sites, or at runtime in dev / SSR. These are wrappers around libraries like Sharp, ImageMagick or Squoosh.

Reviewer: I'd add a bit of nuance here, as it is possible (and probable) for them to be something else

```suggestion
- **Local services** handle image transformations directly at build for static sites, or at runtime in dev / SSR. These are often wrappers around libraries like Sharp, ImageMagick or Squoosh.
```
